akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala

```diff
@@ -661,11 +664,24 @@ final class GraphStageIsland(
     //    } else {
     val props = ActorGraphInterpreter.props(shell)
       .withDispatcher(effectiveSettings.dispatcher)
-    materializer.actorOf(props, islandName)
+    materializer.actorOf(props, fullIslandName)
     //    }
 
   }
 
+  private def fullIslandName: String = {
+    @tailrec def findUsefulName(i: Int): String = {
+      if (i == logics.size) islandName
+      else logics.get(i) match {
+        case _: DownstreamBoundaryStageLogic[_] | _: UpstreamBoundaryStageLogic[_] ⇒
+          findUsefulName(i + 1)
+        case _ ⇒
+          islandName + "-" + logics.get(i).attributes.nameOrDefault()
```

Comment:
so this picks the first logic in the island? Btw, you can make it faster if you just simply store the name in a `var` and fill it in on the first `materializeAtomic` call.